Fail to run gen-certs.sh in mac

In Mac, I got an error when running gen-certs.sh. I tried to uncomment the statements for Mac in gen-certs.sh, but it didn't help.

$ ./gen-certs.sh 
+ gcloud container clusters list '--format=value(name,zone)'
+ grep game-cluster
+ read -r line
+ cluster_setup game-cluster-eu-1 europe-west4-b
+ echo 'Setting up cluster: game-cluster-eu-1 europe-west4-b'
Setting up cluster: game-cluster-eu-1 europe-west4-b
+ gcloud container clusters get-credentials game-cluster-eu-1 --zone europe-west4-b
Fetching cluster endpoint and auth data.
kubeconfig entry generated for game-cluster-eu-1.
++ kubectl config current-context
+ CERT_FOLDER=./certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1
+ mkdir -p ./certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1
+ server_cert game-cluster-eu-1 europe-west4-b
+ echo 'Setting up server certs: game-cluster-eu-1 europe-west4-b'
Setting up server certs: game-cluster-eu-1 europe-west4-b
+ cat
+ kubectl apply -f -
clusterissuer.cert-manager.io/selfsigned unchanged
++ kubectl get services agones-allocator -n agones-system -o 'jsonpath={.status.loadBalancer.ingress[0].ip}'
+ EXTERNAL_IP=34.91.208.135
+ cat
+ kubectl apply -f -
certificate.cert-manager.io/allocator-selfsigned-cert unchanged
+ TLS_CA_FILE=./certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/ca.crt
++ kubectl get secret allocator-tls -n agones-system '-ojsonpath={.data.ca\.crt}'
+ TLS_CA_VALUE=LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUM2ekNDQWRPZ0F3SUJBZ0lRQ1YwOEtPcFc0K0E4MVdvN3k2N0ZVakFOQmdrcWhraUc5dzBCQVFzRkFEQVkKTVJZd0ZBWURWUVFERXcwek5DNDVNUzR5TURndU1UTTFNQjRYRFRJd01EY3dOekl3TURFeE0xb1hEVEl3TVRBdwpOVEl3TURFeE0xb3dHREVXTUJRR0ExVUVBeE1OTXpRdU9URXVNakE0TGpFek5UQ0NBU0l3RFFZSktvWklodmNOCkFRRUJCUUFEZ2dFUEFEQ0NBUW9DZ2dFQkFMM1crZ1J4YUpXbkdaQ0ZHMEtsNHVteS83YjJLWGoxUGZTeFl1Tm0Kc2FJQkJPaEVTZGRnVDRCb3JGVXE0VGZ6WXIxOHZvNXkrSjRXUW0xMEtlSlRqeTlobEdRTzhwYng2V2NtYTZ2Rwp5dVJkMDR2OE5CRFU1THM1S2tOYlh1NWdEemlzUGx5dXJVY25SWGFDc2JTdlJUdkh0dUVBOWowZHpUVy9KU0RnCmpMemdwUFdhbG13aFl5MjUyNFN3UnJGZXpTd3A3WHdHYStZNmJJWWpiYStFV3VLUis5MkRUa1FLM3F3LzJQMW0Kd0taMyt5OUJEMC9CdWJzRHN6M2ViQUk0cWNacGUyYW1YQURSTXhCUmZUb3p5d2VmWTRlZiswTjhjSU9UUUMwawpBRVRTck5hOHcrdHJUY0xuUGp1ZFpTd0JBUVo3V3BaVWNNZ1J6cUd2MzFzOGloMENBd0VBQWFNeE1DOHdEZ1lEClZSMFBBUUgvQkFRREFnV2dNQXdHQTFVZEV3RUIvd1FDTUFBd0R3WURWUjBSQkFnd0JvY0VJbHZRaHpBTkJna3EKaGtpRzl3MEJBUXNGQUFPQ0FRRUFNV1VUdlVjRGZoQXgxVVoxSWNja2IvTWVOSnZFZ2ZtZ2E1amdzRGx2cm15MApvVVBSR2FFdEw1blVFNzgyK3F1aURzK3pmZlRLT1lVeFhLdURaMnY4SWZ6ZGdDdXN2NG01UXh3cTJtcksxKzNqCms5Qk1aQ2VrK2lqQmd1WE9WSk5ZekNjblcxUUg2K0s0a2V6dEs0eWZXbHVlWnpwZnpNa3RvdXpkZDltbnJBUnQKZTk3VExDdVJrb2ViV1VlemJHbkpDSFBkdTR3YUFjd0dHL2F6bXJnZ3B5OVEzc0tWY2JKTEpjaU1iS1h2RUQzaApsemFXWGd1U2x0c1FPbjl1UnZISzlLckJnWDVicWdyRGNTOVBHcnhOayswSjFyQ1Q4TDNlMmpnNFQ4bkJHQ1NxCmh4L3lzWGJWNHljZEtYVTcwbjQvZmxQdVNDeXVncWNhVWlDYUlQa3ZIQT09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
+ echo LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUM2ekNDQWRPZ0F3SUJBZ0lRQ1YwOEtPcFc0K0E4MVdvN3k2N0ZVakFOQmdrcWhraUc5dzBCQVFzRkFEQVkKTVJZd0ZBWURWUVFERXcwek5DNDVNUzR5TURndU1UTTFNQjRYRFRJd01EY3dOekl3TURFeE0xb1hEVEl3TVRBdwpOVEl3TURFeE0xb3dHREVXTUJRR0ExVUVBeE1OTXpRdU9URXVNakE0TGpFek5UQ0NBU0l3RFFZSktvWklodmNOCkFRRUJCUUFEZ2dFUEFEQ0NBUW9DZ2dFQkFMM1crZ1J4YUpXbkdaQ0ZHMEtsNHVteS83YjJLWGoxUGZTeFl1Tm0Kc2FJQkJPaEVTZGRnVDRCb3JGVXE0VGZ6WXIxOHZvNXkrSjRXUW0xMEtlSlRqeTlobEdRTzhwYng2V2NtYTZ2Rwp5dVJkMDR2OE5CRFU1THM1S2tOYlh1NWdEemlzUGx5dXJVY25SWGFDc2JTdlJUdkh0dUVBOWowZHpUVy9KU0RnCmpMemdwUFdhbG13aFl5MjUyNFN3UnJGZXpTd3A3WHdHYStZNmJJWWpiYStFV3VLUis5MkRUa1FLM3F3LzJQMW0Kd0taMyt5OUJEMC9CdWJzRHN6M2ViQUk0cWNacGUyYW1YQURSTXhCUmZUb3p5d2VmWTRlZiswTjhjSU9UUUMwawpBRVRTck5hOHcrdHJUY0xuUGp1ZFpTd0JBUVo3V3BaVWNNZ1J6cUd2MzFzOGloMENBd0VBQWFNeE1DOHdEZ1lEClZSMFBBUUgvQkFRREFnV2dNQXdHQTFVZEV3RUIvd1FDTUFBd0R3WURWUjBSQkFnd0JvY0VJbHZRaHpBTkJna3EKaGtpRzl3MEJBUXNGQUFPQ0FRRUFNV1VUdlVjRGZoQXgxVVoxSWNja2IvTWVOSnZFZ2ZtZ2E1amdzRGx2cm15MApvVVBSR2FFdEw1blVFNzgyK3F1aURzK3pmZlRLT1lVeFhLdURaMnY4SWZ6ZGdDdXN2NG01UXh3cTJtcksxKzNqCms5Qk1aQ2VrK2lqQmd1WE9WSk5ZekNjblcxUUg2K0s0a2V6dEs0eWZXbHVlWnpwZnpNa3RvdXpkZDltbnJBUnQKZTk3VExDdVJrb2ViV1VlemJHbkpDSFBkdTR3YUFjd0dHL2F6bXJnZ3B5OVEzc0tWY2JKTEpjaU1iS1h2RUQzaApsemFXWGd1U2x0c1FPbjl1UnZISzlLckJnWDVicWdyRGNTOVBHcnhOayswSjFyQ1Q4TDNlMmpnNFQ4bkJHQ1NxCmh4L3lzWGJWNHljZEtYVTcwbjQvZmxQdVNDeXVncWNhVWlDYUlQa3ZIQT09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
+ base64 -D
+ kubectl get secret allocator-tls-ca -o json -n agones-system
+ jq '.data["tls-ca.crt"]="LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUM2ekNDQWRPZ0F3SUJBZ0lRQ1YwOEtPcFc0K0E4MVdvN3k2N0ZVakFOQmdrcWhraUc5dzBCQVFzRkFEQVkKTVJZd0ZBWURWUVFERXcwek5DNDVNUzR5TURndU1UTTFNQjRYRFRJd01EY3dOekl3TURFeE0xb1hEVEl3TVRBdwpOVEl3TURFeE0xb3dHREVXTUJRR0ExVUVBeE1OTXpRdU9URXVNakE0TGpFek5UQ0NBU0l3RFFZSktvWklodmNOCkFRRUJCUUFEZ2dFUEFEQ0NBUW9DZ2dFQkFMM1crZ1J4YUpXbkdaQ0ZHMEtsNHVteS83YjJLWGoxUGZTeFl1Tm0Kc2FJQkJPaEVTZGRnVDRCb3JGVXE0VGZ6WXIxOHZvNXkrSjRXUW0xMEtlSlRqeTlobEdRTzhwYng2V2NtYTZ2Rwp5dVJkMDR2OE5CRFU1THM1S2tOYlh1NWdEemlzUGx5dXJVY25SWGFDc2JTdlJUdkh0dUVBOWowZHpUVy9KU0RnCmpMemdwUFdhbG13aFl5MjUyNFN3UnJGZXpTd3A3WHdHYStZNmJJWWpiYStFV3VLUis5MkRUa1FLM3F3LzJQMW0Kd0taMyt5OUJEMC9CdWJzRHN6M2ViQUk0cWNacGUyYW1YQURSTXhCUmZUb3p5d2VmWTRlZiswTjhjSU9UUUMwawpBRVRTck5hOHcrdHJUY0xuUGp1ZFpTd0JBUVo3V3BaVWNNZ1J6cUd2MzFzOGloMENBd0VBQWFNeE1DOHdEZ1lEClZSMFBBUUgvQkFRREFnV2dNQXdHQTFVZEV3RUIvd1FDTUFBd0R3WURWUjBSQkFnd0JvY0VJbHZRaHpBTkJna3EKaGtpRzl3MEJBUXNGQUFPQ0FRRUFNV1VUdlVjRGZoQXgxVVoxSWNja2IvTWVOSnZFZ2ZtZ2E1amdzRGx2cm15MApvVVBSR2FFdEw1blVFNzgyK3F1aURzK3pmZlRLT1lVeFhLdURaMnY4SWZ6ZGdDdXN2NG01UXh3cTJtcksxKzNqCms5Qk1aQ2VrK2lqQmd1WE9WSk5ZekNjblcxUUg2K0s0a2V6dEs0eWZXbHVlWnpwZnpNa3RvdXpkZDltbnJBUnQKZTk3VExDdVJrb2ViV1VlemJHbkpDSFBkdTR3YUFjd0dHL2F6bXJnZ3B5OVEzc0tWY2JKTEpjaU1iS1h2RUQzaApsemFXWGd1U2x0c1FPbjl1UnZISzlLckJnWDVicWdyRGNTOVBHcnhOayswSjFyQ1Q4TDNlMmpnNFQ4bkJHQ1NxCmh4L3lzWGJWNHljZEtYVTcwbjQvZmxQdVNDeXVncWNhVWlDYUlQa3ZIQT09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K"'
+ kubectl apply -f -
secret/allocator-tls-ca configured
+ client_cert game-cluster-eu-1 europe-west4-b
+ echo 'Setting up server certs: game-cluster-eu-1 europe-west4-b'
Setting up server certs: game-cluster-eu-1 europe-west4-b
+ KEY_FILE=./certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/client.key
+ CERT_FILE=./certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/client.crt
+ openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout ./certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/client.key -out ./certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/client.crt -subj /C=/ST=/L=/O=/OU=/CN=
Generating a 2048 bit RSA private key
...........+++
.....+++
writing new private key to './certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/client.key'
-----
end of string encountered while processing type of subject name element #5
problems making Certificate Request

However, after I installed OpenSSL in mac and ran gen-certs.sh, allocate.sh returned the following error in mac. (The allocate.sh script does work in Linux.)

+ go run main.go --ip 34.91.133.2 --namespace default --key /Users/yingxin/go/src/googleforgames/cloud-game-servers-examples/terraform-4-cluster/allocation/certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/client.key --cert /Users/yingxin/go/src/googleforgames/cloud-game-servers-examples/terraform-4-cluster/allocation/certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/client.crt --cacert /Users/yingxin/go/src/googleforgames/cloud-game-servers-examples/terraform-4-cluster/allocation/certs/gke_yingxin-devconsole_europe-west4-b_game-cluster-eu-1/ca.crt --multicluster true
panic: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection closed

goroutine 1 [running]:
main.main()
	/Users/yingxin/go/src/agones.dev/agones/examples/allocator-client/main.go:75 +0x7b3
exit status 2

============================================================
The allocator logs show the following error message:

$ kubectl logs agones-allocator-898464d7b-5hjjp --namespace=agones-system 
...
{"message":"client cert /home/allocator/client-ca/client-ca.crt is installed","severity":"info","source":"main","time":"2020-08-20T06:25:20.270375209Z"}
{"message":"client cert /home/allocator/client-ca/client_trial.crt cannot be installed","severity":"error","source":"main","time":"2020-08-20T06:25:20.270421731Z"}

--- a/terraform-4-cluster/README.md
+++ b/terraform-4-cluster/README.md
@@ -54,6 +54,9 @@ use this script export the environment variable `ALLOCATION_PATH` to point to th
 [allocator sample client](https://agones.dev/site/docs/advanced/multi-cluster-allocation/#allocate-multi-cluster)
 directory.
 
+If there is a failure while running `allocate.sh` give it a few minutes before running the script again.
+The wait is for Game Servers to synchronize resources across clusters.
+
 ## Cleaning up
 
 ```shell script

--- a/terraform-4-cluster/agones-cluster/main.tf
+++ b/terraform-4-cluster/agones-cluster/main.tf
@@ -22,7 +22,7 @@ terraform {
 
 // Create a GKE cluster with the appropriate structure
 module "agones_cluster" {
-  source = "git::https://github.com/googleforgames/agones.git//install/terraform/modules/gke/?ref=release-1.8.0"
+  source = "git::https://github.com/googleforgames/agones.git//install/terraform/modules/gke/?ref=release-1.9.0"
 
   cluster = {
     "name"             = var.name
@@ -36,9 +36,9 @@ module "agones_cluster" {
 
 // Install Agones via Helm
 module "helm_agones" {
-  source = "git::https://github.com/googleforgames/agones.git//install/terraform/modules/helm3/?ref=release-1.8.0"
+  source = "git::https://github.com/googleforgames/agones.git//install/terraform/modules/helm3/?ref=release-1.9.0"
 
-  agones_version         = "1.8.0"
+  agones_version         = "1.9.0"
   values_file            = ""
   chart                  = "agones"
   host                   = module.agones_cluster.host
@@ -81,7 +81,7 @@ resource "helm_release" "cert_manager" {
   force_update = "true"
   repository = data.helm_repository.jetstack.metadata.0.name
   chart = "cert-manager"
-  version = "v0.15.1"
+  version = "v1.0.3"
   timeout = 420
   namespace = "cert-manager"
   create_namespace = true

--- a/terraform-4-cluster/allocation/gen-certs.sh
+++ b/terraform-4-cluster/allocation/gen-certs.sh
@@ -37,7 +37,7 @@ server_cert() {
   #!/bin/bash
   # Create a self-signed ClusterIssuer
 cat <<EOF | kubectl apply -f -
-apiVersion: cert-manager.io/v1alpha2
+apiVersion: cert-manager.io/v1
 kind: ClusterIssuer
 metadata:
   name: selfsigned
@@ -49,7 +49,7 @@ EOF
 
   # Create a Certificate with IP for the allocator-tls secret
 cat <<EOF | kubectl apply -f -
-apiVersion: cert-manager.io/v1alpha2
+apiVersion: cert-manager.io/v1
 kind: Certificate
 metadata:
   name: allocator-selfsigned-cert
@@ -67,10 +67,13 @@ EOF
   # Optional: Store the secret ca.crt in a file to be used by the client for the server authentication
   TLS_CA_FILE=$CERT_FOLDER/ca.crt
   TLS_CA_VALUE=`kubectl get secret allocator-tls -n agones-system -ojsonpath='{.data.ca\.crt}'`
-  echo ${TLS_CA_VALUE} | base64 -d > ${TLS_CA_FILE}
 
-  # In case of MacOS
-  # echo ${TLS_CA_VALUE} | base64 -D > ${TLS_CA_FILE}
+  if [[ "$OSTYPE" == "darwin"* ]]; then
+    # Mac OSX
+    echo ${TLS_CA_VALUE} | base64 -D > ${TLS_CA_FILE}
+  else
+    echo ${TLS_CA_VALUE} | base64 -d > ${TLS_CA_FILE}
+  fi
 
   # Add ca.crt to the allocator-tls-ca Secret
   kubectl get secret allocator-tls-ca -o json -n agones-system | jq '.data["tls-ca.crt"]="'${TLS_CA_VALUE}'"' | kubectl apply -f -
@@ -82,12 +85,14 @@ client_cert() {
   KEY_FILE=$CERT_FOLDER/client.key
   CERT_FILE=$CERT_FOLDER/client.crt
 
-  openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout ${KEY_FILE} -out ${CERT_FILE} -subj "/C=/ST=/L=/O=/OU=/CN="
+  openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout ${KEY_FILE} -out ${CERT_FILE} -subj "/CN=client"
 
-  CERT_FILE_VALUE=`cat ${CERT_FILE} | base64 -w 0`
-
-  # In case of MacOS
-  # CERT_FILE_VALUE=`cat ${CERT_FILE} | base64`
+  if [[ "$OSTYPE" == "darwin"* ]]; then
+    # Mac OSX
+    CERT_FILE_VALUE=`cat ${CERT_FILE} | base64`
+  else
+    CERT_FILE_VALUE=`cat ${CERT_FILE} | base64 -w 0`
+  fi
 
   # white-list client certificate
   kubectl get secret allocator-client-ca -o json -n agones-system | jq '.data["client_trial.crt"]="'${CERT_FILE_VALUE}'"' | kubectl apply -f -
@@ -97,5 +102,9 @@ gcloud container clusters list --format="value(name,zone)" | grep game-cluster |
     cluster_setup $line
     server_cert $line
     client_cert $line
-    kubectl get pods -n agones-system -o=name | grep agones-allocator | xargs kubectl delete -n agones-system
 done
+
+# The following commands are used to trigger Game Servers reconciliation of clusters within realm, instead of waiting for an hour for the changes to have effect. 
+echo " The update realm is to trigger reconciliation between clusters."
+gcloud game servers realms update united-states --update-labels=usage=testing --location=global --no-dry-run >/dev/null
+gcloud game servers realms update europe --update-labels=usage=testing --location=global --no-dry-run >/dev/null

